src/main/scala/is/hail/io/index/IndexReader.scala

```diff
@@ -83,6 +84,88 @@ class IndexReader(hConf: Configuration, path: String, cacheCapacity: Int = 8) ex
     }
   }
 
+  private[io] def binarySearchLowerBound(n: Int, key: Annotation, f: (Int) => Annotation): Int = {
+    var left = 0
+    var right = n
+    while (left < right) {
+      val mid = left + (right - left) / 2
+      if (ordering.lt(f(mid), key))
+        left = mid + 1
+      else
+        right = mid
+    }
+    left // if left < n and A[left] == key, leftmost element that equals key. Otherwise, left is the insertion point of key in A.
+  }
+
+  private[io] def binarySearchUpperBound(n: Int, key: Annotation, f: (Int) => Annotation): Int = {
+    var left = 0
+    var right = n
+    while (left < right) {
+      val mid = left + (right - left) / 2
+      if (ordering.gt(f(mid), key))
+        right = mid
+      else
+        left = mid + 1
+    }
+    left // if left > 0 and A[left - 1] == key, rightmost element that equals key. Otherwise, left is the insertion point of key in A.
+  }
+
+  private def lowerBound(key: Annotation, level: Int, offset: Long): Long = {
+    if (level == 0) {
+      val node = readLeafNode(offset)
+      val idx = binarySearchLowerBound(node.children.length, key, { i => node.children(i).key })
+      node.firstIndex + idx
+    } else {
+      val node = readInternalNode(offset)
+      val children = node.children
+      val n = children.length
+      if (n == 0)
```

Comment:
Yes, there can be. I was thinking about checking for `nKeys == 0` earlier in `queryByKey` and `queryByInterval` to avoid this check, but decided this was better.